### PR TITLE
Add spooky_hash_v2_32 and spooky_hash_v2_64 Presto functions

### PR DESCRIPTION
**Pull request by isadikov**

### CODE REVIEW

**Comment by isadikov** [comment]
@aditi-pandit Can you review when you have time? Thanks.

**Comment by mbasmanova** [comment]
@isadikov Welcome to Velox community. Take some time to review https://github.com/facebookincubator/velox/blob/main/CONTRIBUTING.md and https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md and adjust the code to following the coding standard. Also, please make sure to run the Fuzzer tests for the new functions and write documentation.

**Comment by mbasmanova** [comment]
> SpookyHash requires [changes for big endian platforms](https://burtleburtle.net/bob/hash/spooky.html)

FYI, Velox doesn't support big endian platforms.

**Comment by isadikov** [comment]
Presto does which could potentially result in a correctness issue depending on how the Java library for SpookyHash is written (I could not find the implementation online). I was simply trying to call out the implementation detail.

**Comment by isadikov** [comment]
I have updated the description, thanks @mbasmanova!

File: velox/functions/prestosql/StringFunctions.h [hunk 1]
```diff
@@ -154,6 +154,36 @@ struct HmacSha512Function {
   }
 };
 
+/// spooky_hash_v2_32(varbinary) -> varbinary
+template <typename T>
+struct SpookyHashV2_32Function {
```
Comment (mbasmanova) [inline velox/functions/prestosql/StringFunctions.h:159]: naming: class and struct names should not use underscores

File: velox/functions/prestosql/StringFunctions.h [hunk 2]
```diff
@@ -154,6 +154,36 @@ struct HmacSha512Function {
...
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  template <typename TTo, typename TFrom>
```
Comment (mbasmanova) [inline velox/functions/prestosql/StringFunctions.h:162]: Do we need these template parameters? Aren't they always arg_type<Varbinary>?

File: velox/functions/prestosql/StringFunctions.h (hunk 1 above)
Comment (aditi-pandit) [inline velox/functions/prestosql/StringFunctions.h:159]: Nit : Remove the "_" in the name of this struct and that below.

File: velox/functions/prestosql/StringFunctions.h [hunk 3]
```diff
@@ -154,6 +154,36 @@ struct HmacSha512Function {
...
+  FOLLY_ALWAYS_INLINE void call(TTo& result, const TFrom& input) {
+    uint32_t hash = folly::Endian::swap32(
+        folly::hash::SpookyHashV2::Hash32(input.data(), input.size(), 0));
+    static const auto hashLen = sizeof(int32_t);
```
Comment (mbasmanova) [inline velox/functions/prestosql/StringFunctions.h:166]: naming: kHashLength

File: velox/functions/prestosql/StringFunctions.h [hunk 4]
```diff
@@ -154,6 +154,36 @@ struct HmacSha512Function {
...
+/// spooky_hash_v2_64(varbinary) -> varbinary
+template <typename T>
+struct SpookyHashV2_64Function {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  template <typename TTo, typename TFrom>
+  FOLLY_ALWAYS_INLINE void call(TTo& result, const TFrom& input) {
+    uint64_t hash = folly::Endian::swap64(
+        folly::hash::SpookyHashV2::Hash64(input.data(), input.size(), 0));
+    static const auto hashLen = sizeof(int64_t);
```
Comment (mbasmanova) [inline velox/functions/prestosql/StringFunctions.h:181]: same

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp [hunk 5]
```diff
@@ -1289,6 +1289,28 @@ TEST_F(StringFunctionsTest, HmacSha512) {
   EXPECT_EQ(std::nullopt, hmacSha512(std::nullopt, "velox"));
 }
 
+TEST_F(StringFunctionsTest, SpookyHashV2_32) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/StringFunctionsTest.cpp:1292]: naming: method names should start with lower case and not use underscores

**Review by mbasmanova — COMMENTED** [review]
@isadikov Looks good overall % some nits below. Please, run the fuzzer and report the results.

File: velox/functions/prestosql/StringFunctions.h (hunk 2 above)
Comment (aditi-pandit) [inline velox/functions/prestosql/StringFunctions.h:162]: Why are you using template variables TTo and TFrom when the input and output type of this function can only be Varbinary ? The rest of the functions in this file are using the specific types only.

File: velox/functions/prestosql/StringFunctions.h [hunk 6]
```diff
@@ -154,6 +154,36 @@ struct HmacSha512Function {
...
+  }
+};
+
+/// spooky_hash_v2_64(varbinary) -> varbinary
+template <typename T>
+struct SpookyHashV2_64Function {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  template <typename TTo, typename TFrom>
+  FOLLY_ALWAYS_INLINE void call(TTo& result, const TFrom& input) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/StringFunctions.h:178]: Please add a comment about the endian-ness from the PR description to here instead. Also note that comments should always end with a full-stop.

File: velox/docs/functions/binary.rst [hunk 7]
```diff
@@ -34,6 +34,14 @@ Binary Functions
 
     Computes the HMAC with sha512 of ``binary`` with the given ``key``.
 
+.. function:: spooky_hash_v2_32(binary) -> varbinary
```
Comment (aditi-pandit) [inline velox/docs/functions/binary.rst:37]: Please use the same description as that in https://prestodb.io/docs/current/functions/binary.html

File: velox/docs/functions/binary.rst [hunk 8]
```diff
@@ -34,6 +34,14 @@ Binary Functions
...
+
+    Computes the SpookyHashV2 32-bit hash of ``binary``.
+
+.. function:: spooky_hash_v2_64(binary) -> varbinary
```
Comment (aditi-pandit) [inline velox/docs/functions/binary.rst:41]: We usually maintain the function names in alphabetical order in the files. But it doesn't seem like that is followed in this file.

Please re-order the functions (though in a separate PR).

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp [hunk 9]
```diff
@@ -1289,6 +1289,28 @@ TEST_F(StringFunctionsTest, HmacSha512) {
...
+  const auto spookyHashV2_32 = [&](std::optional<std::string> arg) {
+    return evaluateOnce<std::string, std::string>(
+        "spooky_hash_v2_32(c0)", {arg}, {VARBINARY()});
+  };
+  // Use the same expected value from TestVarbinaryFunctions of presto java
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/StringFunctionsTest.cpp:1297]: Please add some more test cases :
i) Varchar values with special characters : #,@,$,%,^~ etc
ii) Values with numeric characters
iii) Varchar values that are > 12 characters in length. This causes the Varchar to use out of line storage.

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp [hunk 10]
```diff
@@ -1289,6 +1289,28 @@ TEST_F(StringFunctionsTest, HmacSha512) {
...
+  EXPECT_EQ(hexToDec("6BF50919"), spookyHashV2_32(""));
+  EXPECT_EQ(hexToDec("D382E6CA"), spookyHashV2_32("hello"));
+  EXPECT_EQ(std::nullopt, spookyHashV2_32(std::nullopt));
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/StringFunctionsTest.cpp:1300]: Please move the tests for "" and std::nullopt on consecutive lines. These can be followd by tests for valid values.

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp [hunk 11]
```diff
@@ -1289,6 +1289,28 @@ TEST_F(StringFunctionsTest, HmacSha512) {
...
+}
+
+TEST_F(StringFunctionsTest, SpookyHashV2_64) {
+  const auto spookyHashV2_64 = [&](std::optional<std::string> arg) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/StringFunctionsTest.cpp:1304]: Nit -- remove the "_" from the variable name.

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp [hunk 12]
```diff
@@ -1289,6 +1289,28 @@ TEST_F(StringFunctionsTest, HmacSha512) {
...
+    return evaluateOnce<std::string, std::string>(
+        "spooky_hash_v2_64(c0)", {arg}, {VARBINARY()});
+  };
+  // Use the same expected value from TestVarbinaryFunctions of presto java
+  EXPECT_EQ(hexToDec("232706FC6BF50919"), spookyHashV2_64(""));
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/StringFunctionsTest.cpp:1309]: Nit : Rephrase the comment "The result values are obtained from Presto java spookay_hash_v2_64 function."

**Review by aditi-pandit — COMMENTED** [review]
Thanks Ivan for these changes. Have a round of comments.

File: velox/docs/functions/binary.rst (hunk 8 above)
Comment (isadikov) [reply velox/docs/functions/binary.rst]: Yes, I will open a separate PR to fix the order in the file, thanks for the comment!

**Comment by isadikov** [comment]
Thank you for your help, @aditi-pandit! I will follow up on this and post the fuzzer results.

**Comment by mbasmanova** [comment]
@isadikov Sounds like you have resolved the problem with the Fuzzer? Did you get help from @aditi-pandit ?

CC: @bikramSingh91

**Comment by aditi-pandit** [comment]
> > @mbasmanova Would you be able to help me with the fuzzer issue? I am trying the fuzzer on an Apple M1 laptop like this:
> > `./_build/debug/velox/expression/tests/velox_expression_fuzzer_test --batch_size 10 --only spooky_hash_v2_32 --seed 654901148 --duration_sec=10 --logtostderr=1`
> > The fuzzer runs fine for some time but eventually fails with
> > ```
> > zsh: segmentation fault  ./_build/debug/velox/expression/tests/velox_expression_fuzzer_test  10 --only
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I tried other functions such as `sha256`, `week`, `from_unixtime` and they all fail with the above error. Is there a way to work around this problem? Or is it a known issue?
> > @mbasmanova : I helped Ivan run the fuzzer binary in debug mode. He will step through the execution to find the bug.
> 
> @isadikov This issue might have been #3691 where if you use --only flag and have less than 10 functions specified in it, then it crashes at the end of the run while trying to log stats. #3691 has been merged so can you please rebase and try again.

@bikramSingh91 : Yes, indeed that was the crashing line. After we commented it the fuzzer ran successfully.

**Comment by isadikov** [comment]
Thank you everyone for your help with troubleshooting the fuzzer issue. I have rebased my branch and have successfully run the fuzzer with:
```
make debug
./_build/debug/velox/expression/tests/velox_expression_fuzzer_test --only spooky_hash_v2_32,spooky_hash_v2_64 --seed 654 --duration_sec=120 --logtostderr=1
```

Output:
```
I20230116 10:00:36.181054 2787101 ExpressionFuzzer.cpp:877] ==============================> Started iteration 174614 (seed: 3842871813)
I20230116 10:00:36.181321 2787101 ExpressionVerifier.cpp:80] Executing expression: spooky_hash_v2_32(spooky_hash_v2_64(spooky_hash_v2_64(Y8}saSGEDNE``)))
I20230116 10:00:36.181336 2787101 ExpressionVerifier.cpp:32] 0 vectors as input:
E20230116 10:00:36.181459 2787101 Exceptions.h:68] Line: /Users/ivansadikov/velox/velox/expression/ConstantExpr.cpp:219, Function:appendSqlLiteral, Expression:  Type not supported yet: VARBINARY, Source: USER, ErrorCode: UNSUPPORTED
E20230116 10:00:36.181632 2787101 Exceptions.h:68] Line: /Users/ivansadikov/velox/velox/expression/ConstantExpr.cpp:219, Function:appendSqlLiteral, Expression:  Type not supported yet: VARBINARY, Source: USER, ErrorCode: UNSUPPORTED
I20230116 10:00:36.181685 2787101 ExpressionVerifier.cpp:69] All results match.
I20230116 10:00:36.181702 2787101 ExpressionFuzzer.cpp:943] ==============================> Done with iteration 174614
I20230116 10:00:36.181726 2787101 ExpressionFuzzer.cpp:877] ==============================> Started iteration 174615 (seed: 635942518)
I20230116 10:00:36.181748 2787101 ExpressionVerifier.cpp:80] Executing expression: spooky_hash_v2_64(null)
I20230116 10:00:36.181763 2787101 ExpressionVerifier.cpp:32] 0 vectors as input:
I20230116 10:00:36.181882 2787101 ExpressionVerifier.cpp:69] All results match.
I20230116 10:00:36.181897 2787101 ExpressionFuzzer.cpp:943] ==============================> Done with iteration 174615
I20230116 10:00:36.181921 2787101 ExpressionFuzzer.cpp:823] ==============================> Top 2 by number of rows processed
I20230116 10:00:36.181939 2787101 ExpressionFuzzer.cpp:825] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230116 10:00:36.181953 2787101 ExpressionFuzzer.cpp:829] spooky_hash_v2_32 175131 50.11% 4625351
I20230116 10:00:36.181972 2787101 ExpressionFuzzer.cpp:829] spooky_hash_v2_64 174349 49.89% 4574023
I20230116 10:00:36.181985 2787101 ExpressionFuzzer.cpp:835] ==============================> Bottom 2 by number of rows processed
I20230116 10:00:36.182004 2787101 ExpressionFuzzer.cpp:837] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230116 10:00:36.182018 2787101 ExpressionFuzzer.cpp:842] spooky_hash_v2_64 174349 49.89% 4574023
I20230116 10:00:36.182034 2787101 ExpressionFuzzer.cpp:842] spooky_hash_v2_32 175131 50.11% 4625351
I20230116 10:00:36.182047 2787101 ExpressionFuzzer.cpp:855] ==============================> All stats sorted by number of times the function was chosen
I20230116 10:00:36.182061 2787101 ExpressionFuzzer.cpp:857] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230116 10:00:36.182075 2787101 ExpressionFuzzer.cpp:861] spooky_hash_v2_32 175131 50.11% 4625351
I20230116 10:00:36.182090 2787101 ExpressionFuzzer.cpp:861] spooky_hash_v2_64 174349 49.89% 4574023
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
```

File: velox/docs/functions/binary.rst [hunk 13]
```diff
@@ -22,6 +22,14 @@ Binary Functions
 
     Computes the SHA-512 hash of ``binary``.
 
+.. function:: spooky_hash_v2_32(binary) -> varbinary
```
Comment (mbasmanova) [inline velox/docs/functions/binary.rst:25]: In the documentation function should appear in alphabetical order. Please, move these accordingly.

**Review by mbasmanova — APPROVED** [review]
@isadikov Looks good to me % a comment about function order in the docs.

File: velox/docs/functions/binary.rst (hunk 13 above)
Comment (aditi-pandit) [reply velox/docs/functions/binary.rst:25]: @mbasmanova : There are other more valid inconsistencies in this file in the alphabetical order : hmac_sha() functions need to be moved before md5 and xxhash64 needs to be moved later in the file. In a previous review comment I asked @isadikov to fix the ordering in general in a subsequent PR. Hope that is okay.

Comment (mbasmanova) [reply velox/docs/functions/binary.rst:25]: I see. It is fine to fix existing problems in the separate PR, but let's not introduce more problems with the new functions.

Comment (aditi-pandit) [reply velox/docs/functions/binary.rst:25]: This PR doesn't introduce any new problems as the order of these functions is correct relative to the sha... functions above it. Its the rest of the orders that need to be fixed for the inconsistency.

Comment (mbasmanova) [reply velox/docs/functions/binary.rst:25]: Got it. Thank you for clarifying.

**Pull request closed without merging**